File: 36.py
from collections import defaultdict
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        cols = collections.defaultdict(set)
        rows = collections.defaultdict(set)
        subboxes = collections.defaultdict(set)

        for row in range(9):
            for col in range(9):
                num = board[row][col]

                if num == ".":
                    continue
                
                if(num in rows[row] or num in cols[col] or num in subboxes[(row//3, col//3)]):
                    return False
                
                cols[col].add(num)
                rows[row].add(num)
                subboxes[(row//3, col//3)].add(num)
        return True

File: 36.cpp
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        vector<unordered_set<char>> subbox(9);
        vector<unordered_set<char>> cols(9);
        for(int row=0; row < 9; row++){
            unordered_set<char> lines;
            for(int col=0; col<9;col++){
                char num = board[row][col];
                
                //verifica se num é diferente de '.' e esta em algum dos sets. retorna falso
                if(num!='.'){
                    int loc = (row/3)*3 + col/3;
                    if(cols[col].find(num) != cols[col].end() || subbox[loc].find(num) != subbox[loc].end() || lines.find(num) != lines.end()) return false;

                //insere em cada set o num
                lines.insert(num);
                cols[col].insert(num);
                subbox[loc].insert(num);

                }
                
            }
        }
        return true;
    }
};